### PR TITLE
IBX-3201: Added invitation email validation

### PR DESCRIPTION
**Pull request by ViniTou**
https://issues.ibexa.co/browse/IBX-3201
https://issues.ibexa.co/browse/IBX-3361

### CODE REVIEW

File: src/lib/Invitation/Persistence/DoctrineGateway.php [hunk 1]
```diff
@@ -223,7 +227,7 @@ public function updateInvitation(string $hash, InvitationUpdateStruct $updateStr
         $query->where(
             $query->expr()->eq(
                 'hash',
-                $query->createNamedParameter($hash, ParameterType::STRING, ':hash')
+                $query->createNamedParameter($hash, ParameterType::STRING, ':current_hash')
```
Comment (mikadamczyk) [inline src/lib/Invitation/Persistence/DoctrineGateway.php:230]: is `':current_hash'` correct?

File: src/lib/Validator/Constraints/EmailInvitationValidator.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+     * @param string $email The value that should be validated
+     * @param \Symfony\Component\Validator\Constraint $constraint The constraint for the validation
+     */
+    public function validate($email, Constraint $constraint)
+    {
+        try {
+            $this->userService->loadUserByEmail($email);
+            $this->context->addViolation($constraint->message, ['%email%' => $email]);
+        } catch (NotFoundException $exception) {
+        }
```
Comment (mikadamczyk) [inline src/lib/Validator/Constraints/EmailInvitationValidator.php:37]:
```suggestion
        } catch (NotFoundException $exception) {
            // Do nothing
        }
```

File: src/lib/Validator/Constraints/EmailInvitation.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+use JMS\TranslationBundle\Model\Message;
+use JMS\TranslationBundle\Translation\TranslationContainerInterface;
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * @Annotation
+ */
+class EmailInvitation extends Constraint implements TranslationContainerInterface
+{
+    public $message = 'ibexa.user.invitation.user_with_email_exists';
```
Comment (mikadamczyk) [inline src/lib/Validator/Constraints/EmailInvitation.php:20]: Something is wrong here. It is more of a translation key. The same string is used in https://github.com/ibexa/user/pull/46/files#diff-7d3421ef4fd83610224c188cb29b40b95eec2cd666cf97eb74b5563dcb394623R25

File: src/lib/Invitation/Persistence/DoctrineGateway.php (hunk 1 above)
Comment (ViniTou) [reply src/lib/Invitation/Persistence/DoctrineGateway.php:230]: it is as this is named parameter and `hash` is used to provide new `hash` to update

File: src/lib/Validator/Constraints/EmailInvitation.php (hunk 3 above)
Comment (ViniTou) [reply src/lib/Validator/Constraints/EmailInvitation.php:20]: It is translation key. And it is that way as we reuse it in Validator. All of ours Constraints/Validators goes with that pattern.

Comment (mikadamczyk) [reply src/lib/Validator/Constraints/EmailInvitation.php:20]: you could ignore this :)

File: src/lib/Invitation/Persistence/Handler.php [hunk 4]
```diff
@@ -84,7 +84,7 @@ public function invitationExistsForEmail(
 
     public function markAsUsed(string $hash): void
     {
-        $updateStruct = new InvitationUpdateStruct();
+        $updateStruct = new InvitationUpdateStruct(null, null, null);
```
Comment (mikadamczyk) [inline src/lib/Invitation/Persistence/Handler.php:87]:
```suggestion
        $updateStruct = new InvitationUpdateStruct();
```

**Review by mikadamczyk — APPROVED** [review]

File: src/lib/Validator/Constraints/EmailInvitation.php [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * @Annotation
+ */
+class EmailInvitation extends Constraint implements TranslationContainerInterface
+{
+    public string $message = 'ibexa.user.invitation.user_with_email_exists';
+
+    public static function getTranslationMessages()
+    {
+        return [
+            Message::create('ibexa.user.invitation.user_with_email_exists', 'validators')
```
Comment (konradoboza) [inline src/lib/Validator/Constraints/EmailInvitation.php:25]:
```suggestion
            Message::create($message, 'validators')
```

File: src/lib/Validator/Constraints/EmailInvitationValidator.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    public function __construct(UserService $userService)
+    {
+        $this->userService = $userService;
+    }
+
+    /**
+     * Checks if the passed value is valid.
+     *
+     * @param string $email The value that should be validated
+     * @param \Symfony\Component\Validator\Constraint $constraint The constraint for the validation
```
Comment (konradoboza) [inline src/lib/Validator/Constraints/EmailInvitationValidator.php:29]:
```suggestion
```
Doesn't seem to bring much value.

File: src/lib/Validator/Constraints/EmailInvitationValidator.php [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+{
+    private UserService $userService;
+
+    public function __construct(UserService $userService)
+    {
+        $this->userService = $userService;
+    }
+
+    /**
+     * Checks if the passed value is valid.
```
Comment (konradoboza) [inline src/lib/Validator/Constraints/EmailInvitationValidator.php:26]:
```suggestion
```
It seems pretty self-explanatory.

**Review by webhdx — APPROVED** [review]

File: src/lib/Validator/Constraints/EmailInvitation.php (hunk 5 above)
Comment (ViniTou) [reply src/lib/Validator/Constraints/EmailInvitation.php:25]: That's not gonna work, I would have to create property & private const, and that's just seems to much.

**Review by mnocon — APPROVED** [review]

**Pull request merged**